spek-ide-plugin/intellij-common/src/main/kotlin/org/spekframework/intellij/domain/descriptors.kt

```diff
@@ -0,0 +1,226 @@
+package org.spekframework.intellij.domain
+
+import com.intellij.psi.util.PsiTreeUtil
+import org.jetbrains.kotlin.asJava.toLightMethods
+import org.jetbrains.kotlin.idea.refactoring.fqName.getKotlinFqName
+import org.jetbrains.kotlin.idea.refactoring.isAbstract
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.resolve.lazy.data.KtClassInfoUtil
+import org.spekframework.spek2.runtime.scope.Path
+import org.spekframework.spek2.runtime.scope.PathBuilder
+import java.util.concurrent.ConcurrentHashMap
+
+sealed class ScopeDescriptor(val path: Path, val element: KtElement) {
+    class Group(path: Path, element: KtElement, val children: List<ScopeDescriptor>): ScopeDescriptor(path, element) {
+        fun findDescriptorForElement(element: KtElement): ScopeDescriptor? {
+            return findMatching(this, element)
+        }
+
+        private fun findMatching(current: ScopeDescriptor, element: KtElement): ScopeDescriptor? {
+            if (current.element == element) {
+                return current
+            } else if (current is ScopeDescriptor.Group) {
+                for (child in current.children) {
+                    val descriptor = findMatching(child, element)
+                    if (descriptor != null) {
+                        return descriptor
+                    }
+                }
+            }
+            return null
+        }
+    }
+    class Test(path: Path, element: KtElement): ScopeDescriptor(path, element)
+}
+
+private val SPEK_CLASSES = listOf(
+    "org.spekframework.spek2.Spek"
+)
+
+private val SYNONYM_CLASSES = listOf(
+    "org.spekframework.spek2.meta.Synonym"
+)
+
+private val DESCRIPTIONS_CLASSES = listOf(
+    "org.spekframework.spek2.meta.Descriptions"
+)
+
+object ScopeDescriptorCache {
```

Comment:
Can we make this non-static? I'm afraid that over time there'll be some awkward usages of this cache because it's global static variable

Reply:
also maybe extract into its own file

Reply:
👍